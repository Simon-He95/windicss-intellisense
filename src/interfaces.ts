import type { DynamicUtility } from 'windicss/types/interfaces';
import type { Style } from 'windicss/types/utils/style';

export interface Generator {
  variants: {
    [key: string]: () => Style;
  };
  staticUtilities: {
    [key: string]: Style[];
  };
  dynamicUtilities: DynamicUtility;
}
