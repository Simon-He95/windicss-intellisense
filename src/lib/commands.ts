import { commands } from 'vscode';
import type { ExtensionContext } from 'vscode';

export function registerCommands(ctx: ExtensionContext): void {
  ctx.subscriptions.push(
    commands.registerTextEditorCommand("windicss.interpret", (textEditor, textEdit) => {
      console.log(textEditor.document.uri.fsPath);
    })
  );

  ctx.subscriptions.push(
    commands.registerTextEditorCommand("windicss.compile", (textEditor, textEdit) => {
      console.log(textEditor.document.uri.fsPath);
    })
  );

  ctx.subscriptions.push(
    
  );
  
  // commands.registerTextEditorCommand("windicss.sort", (textEditor, textEdit) => {
  //   // console.log(textEditor.document.uri.fsPath);
  //   const document = textEditor.document;
  //   const text = document.getText();
  //   const parser = new HTMLParser(text);
  //   parser.parseClasses().forEach(({start, end, result}) => {
  //     // console.log(result);
  //     const utilities:string[] = [];
  //     const ignored:string[] = [];
  //     const ast = new ClassParser(result).parse();
  //     ast.forEach(obj => {
  //       if (obj.type === 'utility' && typeof obj.content === 'string') {
  //         utilities.push(obj.raw);
  //       } else if (obj.type === 'group') {

  //       } else {
  //         ignored.push(obj.raw);
  //       }
  //     })
  //     console.log(classes);
  //     textEdit.replace(new Range(document.positionAt(start), document.positionAt(end)), '234');
  //   });
  //   // textEditor.document.lineCount

  //   // textEdit.replace(new Range(new Position(0, 0), new Position(textEditor.document.lineCount, textEditor.document.)), '');
  //   // console.log(text);
  // });
}