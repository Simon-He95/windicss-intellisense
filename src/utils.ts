import { MarkdownString } from "vscode";

export function highlightCSS(css?:string): MarkdownString | undefined {
  if (css) {
    return new MarkdownString(`\`\`\`css\n${css}\n\`\`\``);
  }
}

export function flatColors(colors: {[key:string]:string|{[key:string]:string}}) {
  const flatten: {[key:string]:string} = {};
  for (const [key, value] of Object.entries(colors)) {
    if (typeof value === 'string') {
      flatten[key] = value;
    } else {
      for (const [skey, svalue] of Object.entries(value)) {
        flatten[`${key}-${skey}`] = svalue;
      };
    }
  }
  return flatten;
};

export function isColor(className:string, colors: {[key:string]:string}): number[] | undefined {
  for (const [key, value] of Object.entries(colors)) {
    if (className.endsWith('-' + key)) {
      return hex2RGB(value);
    }
  }
}

export function hex2RGB(hex: string): number[] | undefined {
  const RGB_HEX = /^#?(?:([\da-f]{3})[\da-f]?|([\da-f]{6})(?:[\da-f]{2})?)$/i;
  const [, short, long] = String(hex).match(RGB_HEX) || [];

  if (long) {
    const value = Number.parseInt(long, 16);
    return [value >> 16, (value >> 8) & 0xff, value & 0xff];
  } else if (short) {
    return Array.from(short, (s) => Number.parseInt(s, 16)).map(
      (n) => (n << 4) | n
    );
  }
}