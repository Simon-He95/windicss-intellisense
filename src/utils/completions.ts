import { utilities, negative } from './utilities';
import { flatColors } from './index';
import type { Processor } from 'windicss/lib';
import type { colorObject } from 'windicss/types/interfaces';

export interface Attr {
  static: {
    [key:string]: string[]
  },
  color: {
    [key:string]: {
      label: string
      doc: string
    }[]
  },
  dynamic: {
    [key:string]: {
      label: string
      pos: number
    }[]
  }
}

export interface Completion {
  static: string[],
  color: {
    label: string
    doc: string
  }[],
  dynamic: {
    label: string
    pos: number
  }[]
  attr: Attr
}

export function generateCompletions(processor: Processor, colors: colorObject, attributify = true, prefix = '') {
  const completions: Completion = {
    static: [],
    color: [],
    dynamic: [],
    attr: {
      static: {},
      color: {},
      dynamic: {},
    },
  };

  const staticUtilities = processor.resolveStaticUtilities(true);
  // generate normal utilities completions
  for (const [config, list] of Object.entries(utilities)) {
    list.forEach(utility => {
      const mark = utility.search(/\$/);
      if (mark === -1) {
        completions.static.push(utility);
      } else {
        const key = prefix + utility.slice(0, mark - 1);
        const suffix = utility.slice(mark);
        switch (suffix) {
        case '${static}':
          completions.static = completions.static.concat(
            Object.keys(processor.theme(config, {}) as any)
              .map(i => i === 'DEFAULT' ? key : i.charAt(0) === '-' ? `-${key}${i}` : `${key}-${i}`)
          );
          break;
        case '${color}':
          for (const [k, v] of Object.entries(flatColors(processor.theme(config, colors) as colorObject))) {
            completions.color.push({
              label: prefix + `${key}-${k}`,
              doc: v,
            });
          }
          break;
        default:
          completions.dynamic.push({
            label: prefix + utility,
            pos: utility.length - mark,
          });
          if (config in negative) {
            completions.dynamic.push({
              label: prefix + `-${utility}`,
              pos: utility.length + 1 - mark,
            });
          }
          break;
        }
      }
    });
  }

  // generate attributify completions
  const attr: Attr = { static: {}, color: {}, dynamic: {} };

  if (attributify) {

    for (const [key, value] of Object.entries(staticUtilities)) {
      switch (value[0].meta.group) {
      case 'fontStyle':
      case 'fontSmoothing':
      case 'fontVariantNumeric':
        Array.isArray(attr.static['font']) ? attr.static['font'].push(key) : attr.static['font'] = [ key ];
        break;
      case 'textAlign':
        Array.isArray(attr.static['text']) ? attr.static['text'].push(key.replace(/^text-/, '')) : attr.static['text'] = [ key.replace(/^text-/, '') ];
        break;
      case 'veritcalAlign':
      case 'textDecoration':
      case 'textTransform':
      case 'textOverflow':
      case 'whiteSpace':
      case 'wordBreak':
      case 'writingMode':
      case 'writingOrientation':
      case 'hyphens':
        Array.isArray(attr.static['text']) ? attr.static['text'].push(key) : attr.static['text'] = [ key ];
        break;
      }
    }

    for (const utility of completions.static) {
      const { key, body } = split(utility);
      if (key) {
        attr.static[key] = key in attr.static ? [...attr.static[key], body] : [ body ];
      }
    }

    for (const { label, doc } of completions.color) {
      const { key, body } = split(label);
      if (key) {
        const item = { label: body, doc };
        attr.color[key] = key in attr.color ? [...attr.color[key], item] : [ item ];
      }
    }

    for (const { label, pos } of completions.dynamic) {
      const { key, body } = split(label);
      if (key) {
        const item = { label: body, pos };
        attr.dynamic[key] = key in attr.dynamic ? [...attr.dynamic[key], item] : [ item ];
      }
    }
  }

  completions.static.push(...Object.keys(staticUtilities));
  completions.attr = attr;
  return completions;
}

function split(utility: string) {
  const key = utility.match(/[^-]+/)?.[0];
  if (key) {
    if (['tracking', 'leading'].includes(key)) return { key: 'font', body: utility };
  }
  const body = utility.match(/-.+/)?.[0].slice(1) || '~';
  return { key, body };
}
