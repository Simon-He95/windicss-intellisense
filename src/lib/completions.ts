import { negative, utilities } from '../utils/utilities';
import { flatColors, hex2RGB, buildStyle, buildEmptyStyle } from '../utils';
import { Style } from 'windicss/utils/style';
import { patterns, allowAttr } from '../utils/filetypes';
import { languages, Range, Position, CompletionItem, SnippetString, CompletionItemKind } from 'vscode';

import type Extension from './index';
import type { DocumentSelector } from 'vscode';
import type { Processor } from 'windicss/lib';
import type { colorObject } from 'windicss/types/interfaces';

export interface Attr {
  static: {
    [key:string]: string[]
  },
  color: {
    [key:string]: {
      label: string
      doc: string
    }[]
  },
  dynamic: {
    [key:string]: {
      label: string
      pos: number
    }[]
  }
}

export interface Completion {
  static: string[],
  color: {
    label: string
    doc: string
  }[],
  dynamic: {
    label: string
    pos: number
  }[]
  attr: Attr
}

export default class Completions {
  processor: Processor;
  extension: Extension;
  separator: string;
  prefix: string;
  completions: Completion;

  constructor(extension: Extension, processor: Processor) {
    this.processor = processor;
    this.extension = extension;
    this.separator = processor.config('separator', ':') as string;
    this.prefix = processor.config('prefix', '') as string;
    this.completions = this.generateCompletions();
  }

  generateCompletions(attributify = true) {
    const completions: Completion = {
      static: [],
      color: [],
      dynamic: [],
      attr: {
        static: {},
        color: {},
        dynamic: {},
      },
    };

    completions.static.push(...Object.keys(this.processor.resolveStaticUtilities(true)));
    // generate normal utilities completions
    for (const [config, list] of Object.entries(utilities)) {
      list.forEach(utility => {
        const mark = utility.search(/\$/);
        if (mark === -1) {
          completions.static.push(utility);
        } else {
          const prefix = this.prefix + utility.slice(0, mark - 1);
          const suffix = utility.slice(mark);
          switch (suffix) {
          case '${static}':
            completions.static = completions.static.concat(
              Object.keys(this.processor.theme(config, {}) as any)
                .map(i => i === 'DEFAULT' ? prefix : i.charAt(0) === '-' ? `-${prefix}${i}` : `${prefix}-${i}`)
            );
            break;
          case '${color}':
            for (const [k, v] of Object.entries(flatColors(this.processor.theme(config, this.extension.colors) as colorObject))) {
              completions.color.push({
                label: this.prefix + `${prefix}-${k}`,
                doc: v,
              });
            }
            break;
          default:
            completions.dynamic.push({
              label: this.prefix + utility,
              pos: utility.length - mark,
            });
            if (config in negative) {
              completions.dynamic.push({
                label: this.prefix + `-${utility}`,
                pos: utility.length + 1 - mark,
              });
            }
            break;
          }
        }
      });
    }

    // generate attributify completions
    const attr: Attr = { static: {}, color: {}, dynamic: {} };
    if (attributify) {
      for (const utility of completions.static) {
        const { key, body } = split(utility);
        if (key) {
          attr.static[key] = key in attr.static ? [...attr.static[key], body] : [ body ];
        }
      }

      for (const { label, doc } of completions.color) {
        const { key, body } = split(label);
        if (key) {
          const item = { label: body, doc };
          attr.color[key] = key in attr.color ? [...attr.color[key], item] : [ item ];
        }
      }

      for (const { label, pos } of completions.dynamic) {
        const { key, body } = split(label);
        if (key) {
          const item = { label: body, pos };
          attr.dynamic[key] = key in attr.dynamic ? [...attr.dynamic[key], item] : [ item ];
        }
      }
    }

    completions.attr = attr;
    return completions;
  }

  // register suggestions in class = ... | className = ... | @apply ... | sm = ... | hover = ...
  registerUtilities(ext: DocumentSelector, type: string, pattern?: RegExp, enableUtilty = true, enableVariant = true, enableDynamic = true, enableEmmet = false) {
    return languages.registerCompletionItemProvider(
      ext,
      {
        provideCompletionItems: (document, position) => {
          const text = document.getText(new Range(new Position(0, 0), position));

          if ((!pattern || text.match(pattern) === null) && text.match(patterns[type]) === null) {
            const key = text.match(/\S+(?=\s*=\s*["']?[^"']*$)/)?.[0];
            if ((!key) || !(allowAttr(type) && this.extension.isAttrVariant(key))) return [];
          }

          let completions: CompletionItem[] = [];

          if (enableUtilty) {
            completions = completions.concat(
              this.completions.static.map((classItem, index) => {
                const item = new CompletionItem(classItem, CompletionItemKind.Constant);
                item.sortText = '1-' + index.toString().padStart(8, '0');
                return item;
              })
            );
          }

          if (enableVariant) {
            completions = completions.concat(
              Object.keys(this.extension.variants).map((variant, index) => {
                const item = new CompletionItem(variant + this.separator, CompletionItemKind.Module);
                item.detail = variant;
                item.sortText = '2-' + index.toString().padStart(8, '0');
                // register suggestion after select variant
                item.command = {
                  command: 'editor.action.triggerSuggest',
                  title: variant,
                };
                return item;
              })
            ).concat(
              this.completions.color.map(({ label, doc }, index) => {
                const color = new CompletionItem(label, CompletionItemKind.Color);
                color.sortText = '0-' + index.toString().padStart(8, '0');
                color.documentation = doc;
                return color;
              })
            );
          }

          if (enableDynamic) {
            completions = completions.concat(
              this.completions.dynamic.map(({ label, pos }, index) => {
                const item = new CompletionItem(label, CompletionItemKind.Variable);
                item.sortText = '3-' + index.toString().padStart(8, '0');
                item.command = {
                  command: 'cursorMove',
                  arguments: [{
                    to: 'left',
                    select: true,
                    value: pos,
                  }],
                  title: label,
                };
                return item;
              })
            );
          }

          return completions;
        },

        resolveCompletionItem: (item) => {
          switch (item.kind) {
          case CompletionItemKind.Constant:
            item.documentation = buildStyle(this.processor.interpret(item.label).styleSheet);
            break;
          case CompletionItemKind.Module:
            item.documentation = this.buildVariantDoc(item.detail);
            item.detail = undefined;
            break;
          case CompletionItemKind.Variable:
            // TODO
            break;
          case CompletionItemKind.Color:
            const color = (item.documentation || 'currentColor') as string;
            item.documentation = ['transparent', 'currentColor'].includes(color) ? color : `rgb(${hex2RGB(color)?.join(', ')})`;
            item.detail = this.processor.interpret(item.label).styleSheet.build();
            break;
          }
          return item;
        },
      },
      '"',
      '\'',
      ':',
      '!',
      '(',
      ' ',
      ...(enableEmmet ? [ '.' ] : []),
    );
  }

  // register suggestion for bg, text, sm, hover ...
  registerAttrKeys(ext: DocumentSelector, enableUtility = true, enableVariant = true) {
    return languages.registerCompletionItemProvider(
      ext,
      {
        provideCompletionItems: (document, position) => {
          const text = document.getText(new Range(new Position(0, 0), position));
          if (text.match(/(<\w+\s*)[^>]*$/) !== null) {
            if (!text.match(/\S+(?=\s*=\s*["']?[^"']*$)/) || text.match(/<\w+\s+$/)) {
              let completions: CompletionItem[] = [];
              if (enableUtility) {
                completions = completions.concat(
                  Object.keys(this.completions.attr.static).map(label => attrKey(label, CompletionItemKind.Field, 0))
                );
              }
              if (enableVariant) {
                completions = completions.concat(
                  Object.keys(this.extension.variants).map(label => attrKey(label, CompletionItemKind.Module, 1))
                );
              }
              return completions;
            }
          }
          return [];
        },
        resolveCompletionItem: item => {
          switch (item.kind) {
          case CompletionItemKind.Field:
            item.documentation = this.buildAttrDoc(item.label);
            break;
          case CompletionItemKind.Module:
            item.documentation = this.buildVariantDoc(item.label, true);
            break;
          }
          return item;
        },
      },
      ':',
    );
  }

  registerAttrValues(ext: DocumentSelector, enableUtility = true, enableVariant = true, enableDynamic = true) {
    return languages.registerCompletionItemProvider(
      ext,
      {
        provideCompletionItems: (document, position) => {
          const text = document.getText(new Range(new Position(0, 0), position));
          if (text.match(/(<\w+\s*)[^>]*$/) !== null) {
            const key = this.extension.isAttrUtility(text.match(/\S+(?=\s*=\s*["']?[^"']*$)/)?.[0]);
            if (!key) return [];

            let completions: CompletionItem[] = [];
            if (enableUtility) {
              completions = completions.concat(
                this.completions.attr.static[key].map((label, index) => {
                  const item = new CompletionItem(label, CompletionItemKind.Constant);
                  item.detail = key;
                  item.sortText = '1-' + index.toString().padStart(8, '0');
                  return item;
                })
              ).concat(
                key in this.completions.attr.color ? this.completions.attr.color[key].map(({ label, doc }, index) => {
                  const color = new CompletionItem(label, CompletionItemKind.Color);
                  color.sortText = '0-' + index.toString().padStart(8, '0');
                  color.detail = key;
                  color.documentation = doc;
                  return color;
                }) : []
              );
            }

            if (enableVariant) {
              completions = completions.concat(
                Object.keys(this.extension.variants).map((variant, index) => {
                  const item = new CompletionItem(variant + this.separator, CompletionItemKind.Module);
                  item.detail = key + ',' + variant;
                  item.sortText = '2-' + index.toString().padStart(8, '0');
                  item.command = {
                    command: 'editor.action.triggerSuggest',
                    title: variant,
                  };
                  return item;
                })
              );
            }

            if (enableDynamic && key in this.completions.attr.dynamic) {
              completions = completions.concat(
                this.completions.attr.dynamic[key].map(({ label, pos }, index) => {
                  const item = new CompletionItem(label, CompletionItemKind.Variable);
                  item.sortText = '3-' + index.toString().padStart(8, '0');
                  item.command = {
                    command: 'cursorMove',
                    arguments: [{
                      to: 'left',
                      select: true,
                      value: pos,
                    }],
                    title: label,
                  };
                  return item;
                })
              );
            }

            return completions;
          }
        },

        resolveCompletionItem: item => {
          switch (item.kind) {
          case CompletionItemKind.Constant:
            item.documentation = buildStyle(this.processor.attributify({ [item.detail ?? ''] : [ item.label ] }).styleSheet);
            item.detail = undefined;
            break;
          case CompletionItemKind.Module:
            const [attr, variant] = item.detail?.split(',') || [];
            item.documentation = this.buildAttrDoc(attr, variant, this.separator);
            item.detail = undefined;
            break;
          case CompletionItemKind.Variable:
            break;
          case CompletionItemKind.Color:
            const color = (item.documentation || 'currentColor') as string;
            item.documentation = ['transparent', 'currentColor'].includes(color) ? color : `rgb(${hex2RGB(color)?.join(', ')})`;
            item.detail = this.processor.attributify({ [item.detail ?? ''] : [ item.label ] }).styleSheet.build();
            break;
          }
          return item;
        },
      },
      '"',
      '\'',
      ':',
      ' ',
    );
  }

  buildAttrDoc(attr: string, variant?: string, separator?: string) {
    let style;
    if (variant) {
      style = this.extension.variants[variant]();
      style.selector = `[${this.processor?.e(attr)}~="${variant}${separator}&"]`;
    } else {
      style = new Style(`[${this.processor?.e(attr)}~="&"]`);
    }
    return buildEmptyStyle(style);
  }

  buildVariantDoc(variant?: string, attributify = false) {
    if (!variant) return '';
    const style = this.extension.variants[variant]();
    if (attributify) {
      style.selector = `[${this.processor?.e(variant)}~="&"]`;
    } else {
      style.selector = '&';
    }

    return buildEmptyStyle(style);
  }
}

function split(utility: string) {
  const key = utility.match(/[^-]+/)?.[0];
  const body = utility.match(/-.+/)?.[0].slice(1) || '~';
  return { key, body };
}

function attrKey(label: string, kind: CompletionItemKind, order: number) {
  const item = new CompletionItem(label, kind);
  item.sortText = `${order}-` + label;
  item.insertText = new SnippetString(`${label}="$1"`);
  item.command = {
    command: 'editor.action.triggerSuggest',
    title: label,
  };
  return item;
}
