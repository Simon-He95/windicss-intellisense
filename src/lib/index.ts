import jiti from 'jiti';
import Completions from './completions';
import Commands from './commands';
import Hovers from './hovers';
import CodeFolding from './folding';
import Diagnostics from './diagnostics';
import Decorations from './decorations';

import { resolve } from 'path';
import { Processor } from 'windicss/lib';
import { Log } from '../utils/log';
import { getConfig, hex2RGB, flatColors } from '../utils';
import { allowAttr, fileTypes } from '../utils/filetypes';
import { Disposable, workspace, window } from 'vscode';

import type { Attr } from './completions';
import type { ExtensionContext, GlobPattern } from 'vscode';
import type { DictStr, ResolvedVariants, colorObject } from 'windicss/types/interfaces';

export default class Extension {
  _jiti = jiti(__filename);
  ctx: ExtensionContext;
  pattern: GlobPattern;
  processor: Processor | undefined;
  attrs: Attr;
  colors: DictStr;
  variants: ResolvedVariants;
  disposables: Disposable[];
  constructor(ctx: ExtensionContext, pattern: GlobPattern) {
    this.ctx = ctx;
    this.pattern = pattern;
    this.colors = {};
    this.attrs = { static: {}, color: {}, dynamic: {} };
    this.variants = {};
    this.disposables = [];
  }

  async init() {
    try {
      let config;
      const files = await workspace.findFiles(this.pattern, '**â€‹/node_modules/**');
      if (files[0]) {
        const file = files[0].fsPath;
        const path = resolve(file);
        if (path in this._jiti.cache) delete this._jiti.cache[path];
        const exports = this._jiti(path);
        config = exports.__esModule ? exports.default : exports;
        Log.info(`Loading Config File: ${file}`);
      }
      this.processor = new Processor(config);
      this.variants = this.processor.resolveVariants();
      this.colors = flatColors(this.processor.theme('colors', {}) as colorObject);
      this.register();
    } catch (error) {
      Log.warning(error);
    }
  }

  register() {
    const disposables = [
      ...this.registerDecorations(),
      ...this.registerCompletions(),
      ...this.registerDiagnostics(),
      ...this.registerCommands(),
      ...this.registerHovers(),
    ];
    this.ctx.subscriptions.push(...disposables);
    this.disposables.push(...disposables);
  }

  async update() {
    this.disposables.forEach(i => i.dispose());
    this.disposables.length = 0;
    this.register();
  }

  get<T>(key: string) {
    return getConfig<T>(`windicss.${key}`);
  }

  watch() {
    const watcher = workspace.createFileSystemWatcher(this.pattern);
    // Changes configuration should invalidate above cache
    watcher.onDidChange(this.update);

    // This handles the case where the project didn't have config file
    // but was created after VS Code was initialized
    watcher.onDidCreate(this.update);

    // If the config is deleted, utilities&variants should be regenerated
    watcher.onDidDelete(this.update);

    // when change vscode configuration should regenerate disposables
    workspace.onDidChangeConfiguration(this.update, null, this.ctx.subscriptions);
  }

  registerCommands() {
    const commands = new Commands(this.processor);
    const disposables = commands.register(this.ctx);
    if (this.get<boolean>('sortOnSave')) disposables.push(commands.sortOnSave());
    return disposables;
  }

  registerDiagnostics() {
    const diagnostics = new Diagnostics(this.processor);
    const disposables = diagnostics.register();
    return disposables;
  }

  registerCompletions() {
    if (!this.processor) return [];
    const completions = new Completions(this, this.processor);
    this.attrs = completions.completions.attr;
    const disposables: Disposable[] = [];
    const config = {
      enableUtilty: this.get<boolean>('enableUtilityCompletion'),
      enableVariant: this.get<boolean>('enableVariantCompletion'),
      enableDynamic: this.get<boolean>('enableDynamicCompletion'),
      enableEmmet: this.get<boolean>('enableEmmetCompletion'),
      enableAttrUtility: this.get<boolean>('enableAttrUtilityCompletion'),
      enableAttrVariant: this.get<boolean>('enableAttrVariantCompletion'),
    };
    for (const [ext, { type, pattern }] of Object.entries(fileTypes)) {
      disposables.push(completions.registerUtilities(ext, type, pattern, config.enableUtilty, config.enableVariant, config.enableDynamic, config.enableEmmet));
      allowAttr(type) && disposables.push(completions.registerAttrKeys(ext, config.enableAttrUtility, config.enableAttrVariant));
      allowAttr(type) && disposables.push(completions.registerAttrValues(ext, config.enableAttrUtility, config.enableVariant, config.enableDynamic));
    }
    return disposables;
  }

  registerCodeFolding() {
    if (!this.get<boolean>('enableCodeFolding')) return;
    const codeFolding = new CodeFolding();

    window.visibleTextEditors.forEach(editor => codeFolding.create(editor));

    window.onDidChangeTextEditorSelection(() => codeFolding.update(window.activeTextEditor));

    workspace.onDidChangeTextDocument(e => {
      if (window.activeTextEditor && e.document === window.activeTextEditor.document)
        codeFolding.update(window.activeTextEditor);
    }, null, this.ctx.subscriptions);

    window.onDidChangeActiveTextEditor(() => {
      codeFolding.create();
    }, null, this.ctx.subscriptions);

    window.onDidChangeVisibleTextEditors(() => codeFolding.update(window.activeTextEditor));

    workspace.onDidChangeConfiguration(() => {
      if (this.get<boolean>('enableCodeFolding')) {
        codeFolding.create();
      } else {
        codeFolding.remove();
      }
    }, null, this.ctx.subscriptions);
  }

  registerDecorations() {
    if (!this.processor) return [];
    const disposables = [];
    const decoration = new Decorations(this, this.processor);
    for (const [ext] of Object.entries(fileTypes)) {
      disposables.push(decoration.register(ext));
    }
    return disposables;
  }

  registerHovers() {
    if (!this.processor) return [];
    const disposables = [];
    const decoration = new Hovers(this, this.processor);
    for (const [ext] of Object.entries(fileTypes)) {
      disposables.push(decoration.register(ext));
    }
    return disposables;
  }

  // utils
  isAttr(word: string): boolean {
    const lastKey = word.match(/[^:-]+$/)?.[0] || word;
    return (getConfig('windicss.enableAttrVariantCompletion') && lastKey in this.variants) || (getConfig('windicss.enableAttrUtilityCompletion') && lastKey in this.attrs);
  }

  isAttrVariant(word: string): boolean {
    const lastKey = word.match(/[^:-]+$/)?.[0] || word;
    return getConfig('windicss.enableAttrVariantCompletion') && lastKey in this.variants;
  }

  isAttrUtility(word?: string): string | undefined {
    if (!word) return;
    const lastKey = word.match(/[^:-]+$/)?.[0] || word;
    return getConfig('windicss.enableAttrUtilityCompletion') && lastKey in { ...this.attrs.static, ...this.attrs.color, ...this.attrs.dynamic } ? lastKey : undefined;
  }

  isValidColor(utility: string) {
    if (/hex-?(?:([\da-f]{3})[\da-f]?|([\da-f]{6})(?:[\da-f]{2})?)$/.test(utility)) {
      const hex = utility.replace(/^\S*hex-/, '');
      return { color: hex2RGB('#' + hex), key: 'hex-' + hex };
    }
    for (const [key, value] of Object.entries(this.colors)) {
      if (utility.endsWith(key)) {
        return { color: hex2RGB(Array.isArray(value) ? value[0] : value), key };
      }
    }
    return {};
  }
}
