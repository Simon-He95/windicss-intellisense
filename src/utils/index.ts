import { workspace, MarkdownString, Range, Position, DecorationOptions } from 'vscode';
import { ClassParser } from 'windicss/utils/parser';
import { HTMLParser } from './parser';
import { keyOrder } from './order';
import { flatColors as _flatColors } from 'windicss/utils';
import type { Style, StyleSheet } from 'windicss/utils/style';
import type { colorObject, DictStr } from 'windicss/types/interfaces';

export function flatColors(colors: colorObject) {
  colors = _flatColors(colors);
  for (const [key, value] of Object.entries(colors)) {
    if (typeof value === 'function') {
      colors[key] = 'currentColor';
    }
  }
  return colors as DictStr;
}

export function highlightCSS(css?: string): MarkdownString | undefined {
  if (css) {
    return new MarkdownString(`\`\`\`css\n${css}\n\`\`\``);
  }
}

export function hex2RGB(hex: string): number[] | undefined {
  const RGB_HEX = /^#?(?:([\da-f]{3})[\da-f]?|([\da-f]{6})(?:[\da-f]{2})?)$/i;
  const [, short, long] = String(hex).match(RGB_HEX) || [];

  if (long) {
    const value = Number.parseInt(long, 16);
    return [value >> 16, (value >> 8) & 0xff, value & 0xff];
  } else if (short) {
    return Array.from(short, (s) => Number.parseInt(s, 16)).map(
      (n) => (n << 4) | n
    );
  }
}

export function connectList<T>(list: T[][]) {
  return list.reduce((previous, current) => previous.concat(current), []);
}

export async function decorateWithLength(index: number, line: string, length = 25, color = '#AED0A4', text = '...') {
  return new HTMLParser(line).parseClasses().filter(({ result }) => result.length > length).map(({ start, end, result }) => {
    return {
      range: new Range(new Position(index, start + length), new Position(index, end)),
      renderOptions: {
        after: {
          color,
          contentText: text,
        },
      },
      hoverMessage: result.slice(length,),
    };
  });
}

export async function decorateWithCount(index: number, line: string, count = 3, color = '#AED0A4', text = ' ...') {
  const decorations: DecorationOptions[] = [];
  new HTMLParser(line).parseClasses().forEach(({ start, end, result }) => {
    const classes = new ClassParser(result).parse();
    if (classes[count]) {
      decorations.push({
        range: new Range(new Position(index, start + classes[count].start), new Position(index, end)),
        renderOptions: {
          after: {
            color,
            contentText: text,
          },
        },
        hoverMessage: result.slice(classes[count].start,),
      });
    }
  });
  return decorations;
}

export function sortClassNames(classNames: string, variantsMap: { [key: string]: number }) {
  const ast = new ClassParser(classNames).parse();
  return ast.map(({ raw, variants, important }) => {
    const head = variants.join(':') + ':';
    const utility = raw.replace(head, '');
    const key = utility.match(/\w+/);
    const hasDynamicValue = utility.match(/\d+/);
    const offset = variants.map(i => variantsMap[i] * 100).reduce((p, c) => p + c, 0) + (important ? 500 : 0) + (hasDynamicValue ? 25 : 0);
    if (key === null) return { raw, weight: offset };
    return { raw, weight: (keyOrder[key[0]] ?? 300) + offset };
  }).sort((a, b) => a.weight - b.weight).map(i => i.raw).join(' ');
}

export function getConfig<T = any>(key: string): T | undefined {
  return workspace
    .getConfiguration()
    .get<T>(key);
}

export async function setConfig(key: string, value: any, isGlobal = true) {
  return await workspace
    .getConfiguration()
    .update(key, value, isGlobal);
}

export function toggleConfig(key: string) {
  const config = getConfig(key) as boolean;
  setConfig(key, !config);
}

export function rem2px(str?: string) {
  if (!str) return;
  let index = 0;
  const output: string[] = [];

  while (index < str.length) {
    const rem = str.slice(index,).match(/-?[\d.]+rem;/);
    if (!rem || !rem.index) break;
    const px = ` /* ${parseFloat(rem[0].slice(0, -4)) * 16}px */`;
    const end = index + rem.index + rem[0].length;
    output.push(str.slice(index, end));
    output.push(px);
    index = end;
  }
  output.push(str.slice(index,));
  return output.join('');
}

function componentToHex(c: number) {
  const hex = c.toString(16);
  return hex.length === 1 ? '0' + hex : hex;
}

export function rgb2Hex(r: number, g: number, b: number) {
  return '#' + componentToHex(r*255) + componentToHex(g*255) + componentToHex(b*255);
}

export function arrayEqual(array1: unknown[], array2: unknown[]) {
  return array1.length === array2.length && array1.every((value, index) => value === array2[index]);
}

export function buildStyle(styleSheet?: StyleSheet) {
  return styleSheet ? highlightCSS(getConfig('windicss.enableRemToPxPreview') ? rem2px(styleSheet.build()) : styleSheet.build()) : undefined;
}

export function buildEmptyStyle(style: Style) {
  return highlightCSS(style.build().replace('{\n  & {}\n}', '{\n  ...\n}').replace('{}', '{\n  ...\n}').replace('...\n}\n}', '  ...\n  }\n}'));
}
