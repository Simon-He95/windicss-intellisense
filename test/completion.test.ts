import { Processor } from 'windicss/lib';
import { colorObject } from 'windicss/types/interfaces';
import { generateCompletions } from '../src/utils/completions';

const processor = new Processor();
const completions = generateCompletions(processor, processor.theme('colors') as colorObject, true, '');
const attrs = completions.attr;


it('generate completions', () => {
  expect(attrs.static['font']).toMatchSnapshot();
});
