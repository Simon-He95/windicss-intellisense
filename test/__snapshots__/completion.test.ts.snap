// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate completions 1`] = `
Array [
  "antialiased",
  "subpixel-antialiased",
  "italic",
  "not-italic",
  "normal-nums",
  "ordinal",
  "slashed-zero",
  "lining-nums",
  "oldstyle-nums",
  "proportional-nums",
  "tabular-nums",
  "diagonal-fractions",
  "stacked-fractions",
  "sans",
  "serif",
  "mono",
  "thin",
  "extralight",
  "light",
  "normal",
  "medium",
  "semibold",
  "bold",
  "extrabold",
  "black",
  "tracking-tighter",
  "tracking-tight",
  "tracking-normal",
  "tracking-wide",
  "tracking-wider",
  "tracking-widest",
  "tracking-[var(--tracking)]",
  "leading-3",
  "leading-4",
  "leading-5",
  "leading-6",
  "leading-7",
  "leading-8",
  "leading-9",
  "leading-10",
  "leading-none",
  "leading-tight",
  "leading-snug",
  "leading-normal",
  "leading-relaxed",
  "leading-loose",
  "leading-[var(--leading)]",
]
`;
